### PR TITLE
BB-338 Suspend Lifecycle transition to paused location

### PR DESCRIPTION
**Pull request by welansari**
Issue: [BB-338](https://scality.atlassian.net/browse/BB-338)

*Changes :*

- Adapt the same Pause/Resume API routes used for replication and ingestion to also support lifecycle routes
- Lifecycle routes will use mongo to store the location status
- Backbeat API will be the only process modifying the state of locations in MongoDB, hence scheduling a resume job will be handled in Backbeat API contrary to how it works in other extensions.
- The lifecycle bucket processor will be listening to mongo events using change streams and will be pausing and resuming locations based on the modifications to the state applied to MongoDB

**Note for reviewers:** 
Please review per commit as it will make understanding the changes easier.

[BB-338]: https://scality.atlassian.net/browse/BB-338?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js [hunk 1]
```diff
@@ -328,6 +373,129 @@ class LifecycleBucketProcessor {
...
+                return cb();
+            });
+    }
+
+    /**
+     * Initializes a change stream on the metastore collection
+     * @param {function} cb callback
+     * @returns {undefined}
+     */
+    _setMetastoreChangeStream(cb) {
```
Comment (francoisferrand) [inline extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js]: this is not the "metastore" collection (i.e. the collection with the list of buckets), but the "location status" collection

File: extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js [hunk 2]
```diff
@@ -340,6 +508,9 @@ class LifecycleBucketProcessor {
             done => this._setupProducer(done),
             done => this._initKafkaBacklogMetrics(done),
             done => this._setupConsumer(done),
+            done => this._setupMongoClient(done),
+            done => this._initializeLocationStatuses(done),
```
Comment (francoisferrand) [inline extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js]: there is a race condition here:
* Read initial state → paused
* User triggers a resume, which is added to mongo & all...
* Create the change stream → no event received
* Processor is still paused...

we should first setup the change stream, then look for the initial state ; and in the change stream callback we need to be resilient to both:
* being the first event, e.g. happening before we could do the initial state management
* being already in the expected state
(should not be hard, but just to keep in ming)

File: extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js [hunk 3]
```diff
@@ -328,6 +373,129 @@ class LifecycleBucketProcessor {
         });
     }
 
+    /**
+     * Connects to MongoDB using and retreives
+     * the location status collection
+     * @param {function} cb callback
+     * @returns {undefined} undefined
+     */
+    _setupMongoClient(cb) {
```
Comment (francoisferrand) [inline extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js]: it seems to me the whole mongodb change stream / status management could easily be extracted and made reusable : just need to eventually be able to call "resume" and "suspend" functions on the processor (possibly via callbacks)

it will then be easier to port this change to CRR and Ingestion....

File: extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js [hunk 4]
```diff
@@ -328,6 +373,129 @@ class LifecycleBucketProcessor {
...
+                return cb(err);
+            }
+            // connect to metadata DB
+            this._mongoClient = client.db(this._mongoConfig.database, {
+                ignoreUndefined: true,
+            });
+            // creating a collection in case it doesn't exist, this is done as a safety measure
+            // but normally it will be created by backbeat API that gets deployed before the
+            // lifecycle bucket processor.
+            return this._mongoClient.createCollection(locationStatusCollection, err => {
```
Comment (francoisferrand) [inline extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js]: > locationStatusCollection

don't we need to add "filters" in other parts of Backbeat, to ensure this collection is not unexpectedly processed by replication/ingestion/... ?

File: lib/api/constants.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+const constants = {
+    locationStatusCollection: '__locationStatusStore',
```
Comment (francoisferrand) [inline lib/api/constants.js]: does not seem related to "api", i would expect to find this closer to database : is there some "central" place where we already define the name of the metastore or other system collections?

File: lib/api/BackbeatAPI.js [hunk 6]
```diff
@@ -1088,22 +1124,27 @@ class BackbeatAPI {
...
-        let sites;
-        if (details.site === 'all') {
-            sites = details.extensions[service].filter(s => s !== 'all');
-        } else {
-            sites = [details.site];
-        }
-        sites.forEach(site => {
+        const sites = this.getRequestedSites(details);
+        async.eachLimit(sites, 10, (site, next) => {
+            if (service === 'lifecycle') {
```
Comment (francoisferrand) [inline lib/api/BackbeatAPI.js]: this does not look really nice, with this state management intermingled everywhere (which was indeed already the case)

I am wondering if it would not be better to introduce some kind of interface "LocationStatusManager", called by this api : so we could have 2 implementations (redis/zookeeper and mongo), and easily switch one or the other...

That would be much cleaner IMO, but i am wondering if this is worth it (and how long that would take), esp. since we should eventually change this for CRR and ingestion as well... On the positiive side, it would make it easier to keep both codes around, progressively migrate (e.g. just ingestion) and possibly keep the code compatible with RING (where we will not have mongo, in case we want to upgrade RING to backbeat 8.x)

File: lib/api/BackbeatAPI.js [hunk 7]
```diff
@@ -1171,7 +1261,18 @@ class BackbeatAPI {
             schedule.setHours(schedule.getHours() + hours);
         }
 
-        sites.forEach(site => {
+        return async.eachLimit(sites, 10, (site, next) => {
+            if (service === 'lifecycle') {
```
Comment (XinLi0207) [inline lib/api/BackbeatAPI.js]: we have three services here and doing different kinds of things for each, just my personal thinking, maybe a `switch(service) case` logic is more clear?

File: lib/api/constants.js (hunk 5 above)
Comment (welansari) [reply lib/api/constants.js]: We don't have anything like that actually, the `metastore` collection name for example is hardcoded into the mongo client in arsenal

File: extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js (hunk 4 above)
Comment (welansari) [reply extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js]: I think there's no need for that, in change stream implementation we only listen to buckets defined on the metastore, and in the oplog implementation, this would still be ignored as we only replicate proper objects that have a replication status.

File: extensions/lifecycle/tasks/LifecycleTask.js [hunk 8]
```diff
@@ -916,6 +918,11 @@ class LifecycleTask extends BackbeatTask {
                     return next(err, objectMD);
                 }),
             (objectMD, next) => {
+                // No transition if lifecycle events to location were paused
+                if (this.pausedLocations.has(params.site)) {
```
Comment (nicolas2bert) [inline extensions/lifecycle/tasks/LifecycleTask.js]: *perf* What about moving it up as the "first task" to avoid extra GET requests (from _getObjectMD)?

File: extensions/lifecycle/tasks/LifecycleTask.js [hunk 9]
```diff
@@ -916,6 +918,11 @@ class LifecycleTask extends BackbeatTask {
...
+                    return next(errorLocationPaused);
```
Comment (nicolas2bert) [inline extensions/lifecycle/tasks/LifecycleTask.js]: *debug* Should we add a log with the paused location name?

File: extensions/utils/LocationStatusStream.js [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+ */
+/**
+ * Resume a location
+ * @typedef { (locationName: string) => void } ResumeLocationFn
+ */
+
+/**
+ * @class LocationStatusStream
+ *
+ * @classdesc uses change streams to listen to a location statuss
```
Comment (nicolas2bert) [inline extensions/utils/LocationStatusStream.js:20]: *nit* status

File: extensions/utils/LocationStatusStream.js [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+                return cb(err);
+            }
+            // connect to metadata DB
+            this._mongoClient = client.db(this._mongoConfig.database, {
+                ignoreUndefined: true,
+            });
+            // creating a collection in case it doesn't exist, this is done as a safety measure
+            // but normally it will be created by backbeat API that gets deployed before the
+            // lifecycle bucket processor.
+            return this._mongoClient.createCollection(locationStatusCollection, err => {
```
Comment (nicolas2bert) [inline extensions/utils/LocationStatusStream.js]: *nit* No need to explicitly create the MongoDB collection unless specific options/config are needed.

File: extensions/utils/LocationStatusStream.js [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+                // in case the collection already exists, we ignore the error
+                if (err && err.codeName !== 'NamespaceExists') {
+                    this._log.error('Could not create mongo collection', {
+                        collection: locationStatusCollection,
```
Comment (nicolas2bert) [inline extensions/utils/LocationStatusStream.js]: *debug* You could add  the method to this log.

File: extensions/utils/LocationStatusStream.js [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+                return cb();
+            });
+    }
+
+    /**
+     * Initializes a change stream on the location status collection
+     * @param {function} cb callback
+     * @returns {undefined}
+     */
+    _setChangeStream(cb) {
```
Comment (nicolas2bert) [inline extensions/utils/LocationStatusStream.js:118]: *nit* This function is synchronous.

File: lib/util/LocationStatusManager.js [hunk 14]
```diff
@@ -0,0 +1,552 @@
...
+    _listCollectionDocuments(cb) {
+        this._locationStatusColl.find({}, (err, cursor) => {
+            if (err) {
+                this._logger.error('Could not list documents', {
+                    method: 'LocationStatusManager._listCollectionDocuments',
+                    collection: locationStatusCollection,
+                    error: err.message,
+                });
+            }
+            return cursor.toArray(cb);
```
Comment (nicolas2bert) [inline lib/util/LocationStatusManager.js:100]: same

File: lib/util/LocationStatusManager.js [hunk 15]
```diff
@@ -0,0 +1,552 @@
...
+     * @param {Object[]} locations list of locations data from mongo
+     * @param {function} cb callback
+     * @returns {undefined}
+     */
+    _getPreviousLocationStates(locations, cb) {
+        const validLocationNames = Object.keys(validLocations);
+        locations
+            .filter(loc => validLocationNames.includes(loc._id))
+            .forEach(loc => {
+                this._locationStatusStore[loc._id] = new LocationStatus(supportedServices, loc.value);
```
Comment (nicolas2bert) [inline lib/util/LocationStatusManager.js]: *archi* Why is `LocationStatus` in Arsenal? It seems to be "highly" correlated with Backbeat.

File: lib/util/LocationStatusManager.js [hunk 16]
```diff
@@ -0,0 +1,552 @@
...
+        });
+    }
+
+    /**
+     * Add valid mongo locations to local store
+     * @param {Object[]} locations list of locations data from mongo
+     * @param {function} cb callback
+     * @returns {undefined}
+     */
+    _getPreviousLocationStates(locations, cb) {
```
Comment (nicolas2bert) [inline lib/util/LocationStatusManager.js:110]: *nit* This function looks synchronous.

File: lib/util/LocationStatusManager.js [hunk 17]
```diff
@@ -0,0 +1,552 @@
...
+        const invalidLocations = locations
+            .filter(loc => !validLocationNames.includes(loc._id))
+            .map(loc => loc._id);
+        return this._locationStatusColl.deleteMany({
+            _id: {
+                $in: invalidLocations
+            }
+        }, err => {
+            if (err) {
+                this._logger.error('Couldn not delete invalid locations', {
```
Comment (nicolas2bert) [inline lib/util/LocationStatusManager.js]: *nit* Could

File: lib/util/LocationStatusManager.js [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+            }
+            if (reqBody.hours === '' || isNaN(reqBody.hours) ||
+                parseInt(reqBody.hours, 10) <= 0) {
+                return {
+                    error: errors.MalformedPOSTrequest.customizeDescription(
+                        `${msg}: hours must be an integer greater than 0`),
+                };
+            }
+            return { hours: parseInt(reqBody.hours, 10) };
+        } catch (e) {
```
Comment (nicolas2bert) [inline lib/util/LocationStatusManager.js:401]: *debug* should we log the error?

File: lib/util/LocationStatusManager.js [hunk 19]
```diff
@@ -0,0 +1,552 @@
...
+                return next();
+            }
+            const paused = this._locationStatusStore[location].getServicePauseStatus(service);
+            if (paused) {
+                this._deleteResumeJob(location, service);
+                this._locationStatusStore[location].setServiceResumeSchedule(service, null);
+                return this._updateServiceStatusForLocation(location, next);
+            }
+            return next();
+        }, () => {
```
Comment (nicolas2bert) [inline lib/util/LocationStatusManager.js]: What if `_updateServiceStatusForLocation` fails, should we return an error to our user?
Also, might want to make sure it returns an "arsenal type of error".

File: lib/util/LocationStatusManager.js [hunk 20]
```diff
@@ -0,0 +1,552 @@
...
+                this._pushActionToRedis(service, location, actions.pause);
+                return next();
+            }
+            const paused = this._locationStatusStore[location].getServicePauseStatus(service);
+                if (!paused) {
+                    this._locationStatusStore[location].pauseLocation(service);
+                    return this._updateServiceStatusForLocation(location, next);
+                }
+            return next();
+        }, () => {
```
Comment (nicolas2bert) [inline lib/util/LocationStatusManager.js]: same for error.

File: lib/util/LocationStatusManager.js [hunk 21]
```diff
@@ -0,0 +1,552 @@
...
+            const paused = this._locationStatusStore[location].getServicePauseStatus(service);
+            if (paused) {
+                this._locationStatusStore[location].resumeLocation(service, schedule);
+                if (schedule) {
+                    this._scheduleResumeJob(location, service, schedule);
+                }
+                return this._updateServiceStatusForLocation(location, next);
+            }
+            return next();
+        }, () => {
```
Comment (nicolas2bert) [inline lib/util/LocationStatusManager.js]: same for error

File: extensions/lifecycle/tasks/LifecycleTask.js (hunk 9 above)
Comment (welansari) [reply extensions/lifecycle/tasks/LifecycleTask.js]: we already have that in https://github.com/scality/backbeat/blob/375383c1b4651629349b550784cdcb3bdbd8369d/extensions/lifecycle/tasks/LifecycleTask.js#L976-L983

File: lib/util/LocationStatusManager.js (hunk 15 above)
Comment (welansari) [reply lib/util/LocationStatusManager.js]: The initial reason was cause it's a MongoDB document model, which means that it can get accessed by everyone.

But i guess currently the information is only accessed by backbeat and the logic of pause/resume is also very specific to backbeat. I moved the model to backbeat

File: extensions/utils/LocationStatusStream.js (hunk 13 above)
Comment (welansari) [reply extensions/utils/LocationStatusStream.js:118]: The reason i used callbacks for this and all the other synchronous functions was for cleanness of the code, so that they can fit neatly into the async.series/waterfalls

File: tests/unit/api/BackbeatAPI.spec.js [hunk 22]
```diff
@@ -5,15 +5,18 @@ const BackbeatRequest = require('../../../lib/api/BackbeatRequest');
...
     let bbapi;
 
     const destconfig = config.extensions.replication.destination;
     const { site } = destconfig.bootstrapList[0];
     const ingestSite = `${destconfig.bootstrapList[1].site}-ingestion`;
+    const coldSite = Object.keys(locationConfig)
+        .filter(site => locationConfig[site].isCold)[0];
 
-    before(() => {
+    beforeEach(() => {
```
Comment (KillianG) [inline tests/unit/api/BackbeatAPI.spec.js]: Do we really need to create a new `BackbeatAPI` before every test ? Why wasn't it like that before ?

File: lib/api/BackbeatAPI.js [hunk 23]
```diff
@@ -1383,14 +1253,47 @@ class BackbeatAPI {
...
+                {
+                    crr: {
+                        namespace: zookeeperReplicationNamespace,
+                        statePath: zkReplicationStatePath,
+                    },
+                    ingestion: {
+                        namespace: zookeeperIngestionNamespace,
+                        statePath: zkIngestionStatePath,
+                    },
+                },
```
Comment (francoisferrand) [inline lib/api/BackbeatAPI.js]: nit: maybe we should merge the "topics" and "zookeeper" params in a single map:

```suggestion
                {
                    crr: {
                        topic: this._crrTopic,
                        namespace: zookeeperReplicationNamespace,
                        statePath: zkReplicationStatePath,
                    },
                    ingestion: {
                        topic: this._ingestionTopic,
                        namespace: zookeeperIngestionNamespace,
                        statePath: zkIngestionStatePath,
                    },
                },
```

Comment (francoisferrand) [reply lib/api/BackbeatAPI.js]: and also merge the information from `mongoLocationStatusServices`, somehow: e.g. should have an entry to the service even if it is mongo, though in that case it would NOT have the redis topic or zookeeper state.

File: lib/util/LocationStatusManager.js [hunk 24]
```diff
@@ -0,0 +1,583 @@
...
+        this._zkClient = zkClient;
+        this._redis = redis;
+        this._topics = topics;
+        this._zkDetails = zkDetails;
+        this._logger = logger;
+
+        this._locationStatusColl = null;
+
+        this._locationStatusStore = {};
+        this._scheduledResumeJobs = {
```
Comment (francoisferrand) [inline lib/util/LocationStatusManager.js:57]: can this/should this be an attribute of `LocationStatus` ?
(and thus avaialble through the `locationStatusStore` map)

File: lib/util/LocationStatusManager.js [hunk 25]
```diff
@@ -0,0 +1,583 @@
...
+     * @param {string[]} locations location names to pause
+     * @param {Function} cb callback
+     * @returns {undefined}
+     */
+    _getStateDetails(service, locations, cb) {
+        if (this._mongoServices.includes(service)) {
+            const selectedLocations = {};
+            locations.forEach(loc => {
+                selectedLocations[loc] = this._locationStatusStore[loc];
+            });
```
Comment (francoisferrand) [inline lib/util/LocationStatusManager.js]: this should do the mapping done in the later functions (`.paused = .getServicePauseStatus(service)` etc, c.f. lines 315, 342...), to complete the abstraction :

```suggestion
            locations.forEach(loc => {
                selectedLocations[loc] = this._locationStatusStore[loc].getValue()[service]
            });
```

(or better introduce a `LocationStatus.getService()` function 😉)

File: tests/unit/lib/util/LocationStatusManager.spec.js [hunk 26]
```diff
@@ -0,0 +1,366 @@
...
+            },
+            fakeLogger,
+        );
+    });
+
+    afterEach(() => {
+        sinon.restore();
+    });
+
+    it('pauseService:: should pause selected service', done => {
```
Comment (francoisferrand) [inline tests/unit/lib/util/LocationStatusManager.spec.js]: nit: `pauseService::` should be in a describe block instead, to keep the "flow" of describing the tests

**Review by francoisferrand — APPROVED** [review]

File: lib/util/LocationStatusManager.js (hunk 24 above)
Comment (welansari) [reply lib/util/LocationStatusManager.js:57]: `_scheduledResumeJobs` contains the actual scheduled jobs for all services (that use both mongo and zookeeper), this can't be put in the `LocationStatus` as it'll only work for mongo services

**Review by KillianG — APPROVED** [review]
